tests/suites/test_suite_ssl.function

```diff
@@ -1347,6 +1402,136 @@ static int ssl_populate_session( mbedtls_ssl_session *session,
+ * message was sent in the correct number of fragments.
+ *
+ * /p ssl_1 and /p ssl_2    Endpoints represented by mbedtls_ssl_context. Both
+ *                          of them must be initialized and connected beforehand.
+ * /p msg_len_1 and /p msg_len_2 specify the size of the message to send.
+ * /p expected_fragments_1 and /p expected_fragments_2 determine in how many
+ *                          fragments the message should be sent.
+ *      expected_fragments is 0: can be used for DTLS testing while the message
+ *                          size is larger than MFL
+ *      expected_fragments is 1 - can be used for TLS/DTLS testing while the
+ *                          message size is below MFL
+ *      expected_fragments > 1  - can be used for TLS testing while the message
+ *                          size is larger than MFL
+ *
+ * \retval  0 on success, otherwise error code.
+ */
+int mbedtls_exchange_data( mbedtls_ssl_context *ssl_1,
+                           int msg_len_1, const int expected_fragments_1,
+                           mbedtls_ssl_context *ssl_2,
+                           int msg_len_2, const int expected_fragments_2 )
+{
+    unsigned char *msg_buf_1 = malloc( msg_len_1 );
+    unsigned char *msg_buf_2 = malloc( msg_len_2 );
+    unsigned char *in_buf_1  = malloc( msg_len_2 );
+    unsigned char *in_buf_2  = malloc( msg_len_1 );
+    int msg_type, ret = -1;
+
+    /* Perform this test with two message types. At first use a message
+     * consisting of only 0x00 for the client and only 0xFF for the server.
+     * At the second time use message with generated data */
+    for( msg_type = 0; msg_type < 2; msg_type++ )
+    {
+        int written_1 = 0;
+        int written_2 = 0;
+        int read_1 = 0;
+        int read_2 = 0;
+        int fragments_1 = 0;
+        int fragments_2 = 0;
+
+        if( msg_type == 0 )
+        {
+            memset( msg_buf_1, 0x00, msg_len_1 );
+            memset( msg_buf_2, 0xff, msg_len_2 );
+        }
+        else
+        {
+            int i, j = 0;
+            for( i = 0; i < msg_len_1; i++ )
+                msg_buf_1[i] = j++ & 0xFF;
+            for( i = 0; i < msg_len_2; i++ )
+                msg_buf_2[i] = ( j -= 5 ) & 0xFF;
+        }
+
+        while( read_1 < msg_len_2 || read_2 < msg_len_1 )
+        {
+            /* ssl_1 sending */
+            if( msg_len_1 > written_1 )
+            {
+                ret = mbedtls_ssl_write_fragment( ssl_1, msg_buf_1,
+                                                  msg_len_1, &written_1,
+                                                  &fragments_1,
+                                                  expected_fragments_1 );
+                if( expected_fragments_1 == 0 )
+                {
+                    TEST_ASSERT( ret == MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+                    msg_len_1 = 0;
+                }
+                else
+                {
+                    TEST_ASSERT( ret == 0 );
+                }
```

Comment:
Could we get away with only checking that `ret == 0` here, like we do for reading? Seems we need to update `msg_len_1` sometimes, but it also seems we are duplicating TEST_ASSERTs

Reply:
I do not really see how we could get away that checking. When `expected_fragments_1 == 0` only MBEDTLS_ERR_SSL_BAD_INPUT_DATA is allowed to pass the test, otherwise only `ret == 0`. Would you like to put that both checking into one TEST_ASSERT?

Reply:
Aren't we already checking that MBEDTLS_ERR_SSL_BAD_INPUT_DATA was OK in the function we just called? Perhaps we shouldn't do the checking in the child function and only do it here instead.

Reply:
Testing MBEDTLS_ERR_SSL_BAD_INPUT_DATA here it is required because mbedtls_ssl_write_fragment could also return -1. I can remove TEST_ASSERT from the child function and the test should still work properly, but it made me wonder that we would not detect a potential error as soon as possible. What do you think? Should I remove MBEDTLS_ERR_SSL_BAD_INPUT_DATA from the child function?

Reply:
Making the child function just perform the operation, and having the caller interpret the results seems best to me. It avoids duplication of the check. And also, the caller is in the best position to interpret the results because it is the place where all the logic to run the test is at; the parent knows what is right and what is wrong, and the child just does as it's told.

Reply:
Thanks for that description. I removed checking in the child function and changed the comment